include/openmc/boundary_condition.h

```diff
@@ -16,14 +19,47 @@ class Surface;
 class BoundaryCondition {
 public:
   //! Perform tracking operations for a particle that strikes the boundary.
+  //! Handle the boundary albedo and modify the incident particle's weight.
   //! \param p The particle that struck the boundary.  This class is not meant
   //!   to directly modify anything about the particle, but it will do so
   //!   indirectly by calling the particle's appropriate cross_*_bc function.
   //! \param surf The specific surface on the boundary the particle struck.
-  virtual void handle_particle(Particle& p, const Surface& surf) const = 0;
+  virtual void handle_particle(Particle& p, const Surface& surf) const
+  {
+    if (!has_albedo_)
+      return;
+    // Save incident particle's weight
+    double initial_wgt = p.wgt();
+    // Treat the lost weight fraction as leakage, similar to VacuumBC.
+    // This ensures the lost weight is tallied properly.
+    p.wgt() *= (1.0 - albedo_);
+    p.cross_vacuum_bc(surf);
+    // Reassign the remaining weight and pass particle to BaseBC
+    p.wgt() = initial_wgt * albedo_;
+  };
 
   //! Return a string classification of this BC.
   virtual std::string type() const = 0;
+
+  //! Write albedo data of this BC to hdf5.
+  void to_hdf5(hid_t surf_group) const
+  {
+    if (std::abs(albedo_ - 1.0) <= FP_PRECISION) {
+      write_string(surf_group, "albedo", fmt::format("{}", albedo_), false);
+    }
+  };
+
+  //! Set albedo of this BC.
+  void set_albedo(double albedo)
+  {
+    // Flag that this BC has a weight multiplier for incident particles
+    has_albedo_ = true;
+    albedo_ = albedo;
+  }
+
+private:
+  double albedo_ = 1.0;
+  bool has_albedo_ = false;
```

Comment:
I avoided that because comparing floating points like that might lead to undefined behaviour. Other places in the code do floating point comparison using abs(a - b) < minimum float value logic. I can do that but it will translate to more instructions. Given this check will return every time a particle is processed for any BC, I thought to just use an extra byte in the BC class and do it the way I've done it. Are there drawbacks to my approach that don't justify the optimisation I'm doing?

Reply:
While it is true that floating point comparison is generally unreliable, it should be safe in this particular case because 1) the number 1.0 has an exact representation and 2) we are setting the default value using the same constant against which we are comparing. Of course, it wouldn't be reasonable to expect the result of some mathematical expression to be exactly 1.0 (e.g., `3*0.3 + 0.1`), but for just comparing against a constant I don't see how it could run into problems.

Now, that being said, a safer alternative that would avoid the `!=` comparison is to set the default value to `-1.0` as a "null" value, and then have `bool has_albedo() const { return albedo_ > 0.0; }`

Reply:
That is a sensible solution, will implement that instead.